### PR TITLE
Make the Reflectometry GUI process runs asynchronously

### LINKED ISSUE #19480: Make algorithm execution in GenericDataProcessingPresenter async
The algorithm execution in `GenericDataProcessingPresenter` is not executed asynchronously, which causes Mantid to freeze. It has to be operated async.

### PR DESCRIPTION
**Pull request by PranavBahuguna**
Previously the algorithm execution in the GenericDataProcessorPresenter was not executed asynchronously, which causes Mantid to freeze. This PR allows execution to be performed asynchronously as well as introduce a pause button for halting processing.

**To test:**

- Code review
- Ensure unit tests run are sensible.
- Perform functional testing e.g.:
  - Check that you can operate other parts of the interface while processing runs.
  - Check that window cannot be closed without pausing processing first.
  - Check that only exclusively are the `Process` and `Pause` buttons enabled/disabled, not both.
  - Ensure when pausing processing that the last row/group are processed before fully stopping.
  - Ensure that there is no change in the actual output workspaces after processing from the old interface.

There are probably many ways I haven't thought of to break the interface or do things that have unexpected side effects. Notify me of anything that feels strange to you and I will look into it.

A simple way of obtaining runs to reduce with is to login to the ICAT catalog, set current instrument to `POLREF`, enter id `1520249` and transfer them over.

Fixes #19480.

*Release notes have been updated.*

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [x] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [x] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [x] Has the relevant documentation been added/updated?
- [x] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: MantidQt/MantidWidgets/src/DataProcessorUI/GenericDataProcessorPresenter.cpp [hunk 1]
```diff
@@ -187,71 +187,193 @@ void GenericDataProcessorPresenter::acceptViews(
...
+    // Reduce first row
+    int groupIndex = m_gqueue.front().first;
+    auto &rqueue = m_gqueue.front().second;
+    m_rowItem = rqueue.front();
+    rqueue.pop();
+    // Set next action flag
+    m_nextActionFlag = GenericDataProcessorPresenter::ReduceRowFlag;
+
+    // Prepare a new thread
+    auto *worker = new GenericDataProcessorPresenterRowReducerWorker(
```
Comment (AntonPiccardoSelg) [inline MantidQt/MantidWidgets/src/DataProcessorUI/GenericDataProcessorPresenter.cpp]: See above

**Review by AntonPiccardoSelg — CHANGES_REQUESTED** [review]

File: MantidQt/CustomInterfaces/inc/MantidQtCustomInterfaces/IReflPresenter.h [hunk 2]
```diff
@@ -64,7 +64,8 @@ class IReflPresenter {
     PlotRowFlag,
     PlotGroupFlag,
     ExpandAllGroupsFlag,
-    CollapseAllGroupsFlag
+    CollapseAllGroupsFlag,
+    PauseFlag
```
Comment (PranavBahuguna) [inline MantidQt/CustomInterfaces/inc/MantidQtCustomInterfaces/IReflPresenter.h]: It can be resumed, so `pause` seems appropriate here.

**Review by AntonPiccardoSelg — APPROVED** [review]

**Comment by AntonPiccardoSelg** [comment]
Tested with all of the instructions. The functionality works as expected. The GUI does not freeze any longer and the reduction can be paused and restarted. The outputs are still the same as in the old reduction. Please :shipit:

**Comment by AntonPiccardoSelg** [comment]
@mantidproject/gatekeepers I just wanted to bump this PR, since other pieces of work rely on it.

**Pull request merged**